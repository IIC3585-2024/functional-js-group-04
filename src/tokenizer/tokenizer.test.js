import tokenize from "./tokenizer"

test('tokenizer should be a function', () => {
    expect(typeof tokenize).toBe('function');
});

// CommonMark Spec Blocks

describe("thematic break", () => {
    test('should return the thematic_break token', () => {
        const src = '***\n---\n___\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
});

describe("headings", () => {
    test('should return the heading token', () => {
        const src = '# foo\n## foo\n### foo\n#### foo\n##### foo\n###### foo\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
});

describe("indented code block", () => {
    test('should return the indented_code_block token', () => {
        const src = '    a simple\n      indented code block\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
});

describe("fenced code block", () => {
    test('should return the fenced_code_block token', () => {
        const src = '```javascript\nfunction foo(x) {\n  return 3\n}\n```';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
});

describe("paragraph", () => {
    test('should return the paragraph token', () => {
        const src = 'foo\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test('should return one token on multiple lines', () => {
        const src = 'foo\nbar\nbaz\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
});

describe("blank line", () => {
    test('should return the blank_line token', () => {
        const src = '\n';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
})

describe("inline text", () => {
    test("bold", () => {
        const src = '**foo**';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("italic", () => {
        const src = '*foo*';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("code span", () => {
        const src = '`` foo `bar` baz ``';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("link", () => {
        const src = '[foo](bar "baz")';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
    test("image", () => {
        const src = 'Foo ![foo](/url "title")';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("text_inline", () => {
        const src = 'foo';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("char_inline", () => {
        const src = '*';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("bold and italic", () => {
        const src = '***foo bar***';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("bold and italic and text_inline", () => {
        const src = '***bold and italic***text';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("inline unmatched asterisks", () => {
        const src = '***abc**';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });

    test("link", () => {
        const src = '[**foo**](bar)';
        const tokens = tokenize(src);
        expect(tokens).toMatchSnapshot();
    });
})
