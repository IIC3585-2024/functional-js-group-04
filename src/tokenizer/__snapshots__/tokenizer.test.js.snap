// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blank line should return the blank_line token 1`] = `
[
  {
    "raw": "
",
    "type": "blank_line",
  },
]
`;

exports[`fenced code block should return the fenced_code_block token 1`] = `
[
  {
    "raw": "\`\`\`javascript
function foo(x) {
  return 3
}
\`\`\`",
    "text": "function foo(x) {
  return 3
}",
    "type": "fenced_code_block",
  },
]
`;

exports[`headings should return the heading token 1`] = `
[
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 1,
    "raw": "# foo
",
    "text": "foo",
    "type": "heading",
  },
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 2,
    "raw": "## foo
",
    "text": "foo",
    "type": "heading",
  },
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 3,
    "raw": "### foo
",
    "text": "foo",
    "type": "heading",
  },
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 4,
    "raw": "#### foo
",
    "text": "foo",
    "type": "heading",
  },
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 5,
    "raw": "##### foo
",
    "text": "foo",
    "type": "heading",
  },
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "level": 6,
    "raw": "###### foo
",
    "text": "foo",
    "type": "heading",
  },
]
`;

exports[`indented code block should return the indented_code_block token 1`] = `
[
  {
    "raw": "    a simple
      indented code block
",
    "text": "a simple
  indented code block",
    "type": "indented_code_block",
  },
]
`;

exports[`inline text bold 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "raw": "foo",
            "text": "foo",
            "type": "text_inline",
          },
        ],
        "raw": "**foo**",
        "text": "foo",
        "type": "bold",
      },
    ],
    "raw": "**foo**",
    "text": "**foo**",
    "type": "paragraph",
  },
]
`;

exports[`inline text bold and italic 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "raw": "foo bar",
                "text": "foo bar",
                "type": "text_inline",
              },
            ],
            "raw": "*foo bar*",
            "text": "foo bar",
            "type": "italic",
          },
        ],
        "raw": "***foo bar***",
        "text": "*foo bar*",
        "type": "bold",
      },
    ],
    "raw": "***foo bar***",
    "text": "***foo bar***",
    "type": "paragraph",
  },
]
`;

exports[`inline text bold and italic and text_inline 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "raw": "bold and italic",
                "text": "bold and italic",
                "type": "text_inline",
              },
            ],
            "raw": "*bold and italic*",
            "text": "bold and italic",
            "type": "italic",
          },
        ],
        "raw": "***bold and italic***",
        "text": "*bold and italic*",
        "type": "bold",
      },
      {
        "raw": "text",
        "text": "text",
        "type": "text_inline",
      },
    ],
    "raw": "***bold and italic***text",
    "text": "***bold and italic***text",
    "type": "paragraph",
  },
]
`;

exports[`inline text char_inline 1`] = `
[
  {
    "children": [
      {
        "raw": "*",
        "text": "*",
        "type": "text_inline",
      },
    ],
    "raw": "*",
    "text": "*",
    "type": "paragraph",
  },
]
`;

exports[`inline text code span 1`] = `
[
  {
    "children": [
      {
        "raw": "\`\` foo \`bar\` baz \`\`",
        "text": "foo \`bar\` baz",
        "type": "code_span",
      },
    ],
    "raw": "\`\` foo \`bar\` baz \`\`",
    "text": "\`\` foo \`bar\` baz \`\`",
    "type": "paragraph",
  },
]
`;

exports[`inline text image 1`] = `
[
  {
    "children": [
      {
        "raw": "Foo ",
        "text": "Foo ",
        "type": "text_inline",
      },
      {
        "alt": "foo",
        "children": [],
        "raw": "![foo](/url "title")",
        "src": "/url",
        "title": "title",
        "type": "image",
      },
    ],
    "raw": "Foo ![foo](/url "title")",
    "text": "Foo ![foo](/url "title")",
    "type": "paragraph",
  },
]
`;

exports[`inline text inline unmatched asterisks 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "raw": "*abc",
            "text": "*abc",
            "type": "text_inline",
          },
        ],
        "raw": "***abc**",
        "text": "*abc",
        "type": "bold",
      },
    ],
    "raw": "***abc**",
    "text": "***abc**",
    "type": "paragraph",
  },
]
`;

exports[`inline text italic 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "raw": "foo",
            "text": "foo",
            "type": "text_inline",
          },
        ],
        "raw": "*foo*",
        "text": "foo",
        "type": "italic",
      },
    ],
    "raw": "*foo*",
    "text": "*foo*",
    "type": "paragraph",
  },
]
`;

exports[`inline text link 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "raw": "foo",
            "text": "foo",
            "type": "text_inline",
          },
        ],
        "href": "bar",
        "raw": "[foo](bar "baz")",
        "text": "foo",
        "title": "baz",
        "type": "link",
      },
    ],
    "raw": "[foo](bar "baz")",
    "text": "[foo](bar "baz")",
    "type": "paragraph",
  },
]
`;

exports[`inline text link 2`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "raw": "foo",
                "text": "foo",
                "type": "text_inline",
              },
            ],
            "raw": "**foo**",
            "text": "foo",
            "type": "bold",
          },
        ],
        "href": "bar",
        "raw": "[**foo**](bar)",
        "text": "**foo**",
        "title": null,
        "type": "link",
      },
    ],
    "raw": "[**foo**](bar)",
    "text": "[**foo**](bar)",
    "type": "paragraph",
  },
]
`;

exports[`inline text text_inline 1`] = `
[
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "raw": "foo",
    "text": "foo",
    "type": "paragraph",
  },
]
`;

exports[`paragraph should return one token on multiple lines 1`] = `
[
  {
    "children": [
      {
        "raw": "foo
bar
baz",
        "text": "foo
bar
baz",
        "type": "text_inline",
      },
    ],
    "raw": "foo
bar
baz
",
    "text": "foo
bar
baz",
    "type": "paragraph",
  },
]
`;

exports[`paragraph should return the paragraph token 1`] = `
[
  {
    "children": [
      {
        "raw": "foo",
        "text": "foo",
        "type": "text_inline",
      },
    ],
    "raw": "foo
",
    "text": "foo",
    "type": "paragraph",
  },
]
`;

exports[`thematic break should return the thematic_break token 1`] = `
[
  {
    "raw": "***
",
    "type": "thematic_break",
  },
  {
    "raw": "---
",
    "type": "thematic_break",
  },
  {
    "raw": "___
",
    "type": "thematic_break",
  },
]
`;
